### DISCUSSION TITLE
Adding Future Regressor Error

### DISCUSSION THREAD

**Discussion by Msaleh87** [ourownstory/neural_prophet#799] [Q&A - get help using NeuralProphet] [answered]
### Discussed in https://github.com/ourownstory/neural_prophet/discussions/797

<div type='discussions-op-text'>

<sup>Originally posted by **Msaleh87** October  3, 2022</sup>
So I prepared my data as shown below:
![Screenshot 2022-10-03 165322](https://user-images.githubusercontent.com/42920113/193680423-88e3a373-898d-41fd-a557-bd5f244ee402.png)

This is my code, L12 is the dataset you see in the pic above:

```
L12M = L_prophet_5min_T[0:(12*30*24*12)] # the dataset show in the above pic
m = NeuralProphet(n_forecasts=(2*24*12),
    n_lags=(16*24*12),
    yearly_seasonality="auto",
    weekly_seasonality="auto",  
    daily_seasonality="auto",
    growth="off", seasonality_reg= 0.5, optimizer = 'AdamW',
    learning_rate = 0.1, n_changepoints="auto", changepoints_range = "auto",  batch_size = 16 ) 

m.add_country_holidays("US", mode="additive") 
m = m.add_future_regressor(name='A')
m = m.add_future_regressor(name='B')
m = m.add_future_regressor(name='C')
metrics = m.fit(L12M, freq="5min") 
m.highlight_nth_step_ahead_of_each_forecast(step_number=m.n_forecasts)

# the future values of A, B, and C
future_regressors_df = pd.DataFrame(data={'A': L_prophet_5min_T['A'][(12*30*24*12):(12*30*24*12)+576], 
                                          'B': L_prophet_5min_T['B'][(12*30*24*12):(12*30*24*12)+576],
                                          'C': L_prophet_5min_T['C'][(12*30*24*12):(12*30*24*12)+576]})
```
The model trains and saves fine, then when I try to forecast using this code, I get an error:

```
P_horizon = 576
future = m.make_future_dataframe(df = L12M, periods = P_horizon, n_historic_predictions=True) 
forecast1 = m.predict(df=future)
```
![Screenshot 2022-10-03 170128](https://user-images.githubusercontent.com/42920113/193681742-af08552c-5df4-4586-9a2c-b411ac3c076c.png)

Should I add an argument for the future regressors inside the forecast?!</div>

**Comment by ourownstory** [comment, marked as answer]
> Should I add an argument for the future regressors inside the forecast?!

Their future values should be added in `make_future_dataframe`, see:
https://github.com/ourownstory/neural_prophet/blob/3bf033659acc3d738e2d564da5a81c3ce4951d4c/neuralprophet/forecaster.py#L1164

Hope this helps

**Reply by Msaleh87** [reply]
Thanks @ourownstory, it worked

So I needed to have the values of the future regressors with index starts from 0 and I needed to `regressor_df` to `make_future_dataframe` as you said. I will leave the code here for future reference:

**This is how I sat up the future_regressors_df** 
```
future_regressors_df = pd.DataFrame(data={'A': L_prophet_5min_T['A'][(12*30*24*12):(12*30*24*12)+(288*2)], 
                                          'B': L_prophet_5min_T['B'][(12*30*24*12):(12*30*24*12)+(288*2)],
                                          'C': L_prophet_5min_T['C'][(12*30*24*12):(12*30*24*12)+(288*2)]})
# To make sure the indexing starts from 0, I did this:
future_regressors_df = future_regressors_df.reset_index() 
future_regressors_df = future_regressors_df.drop(['index'], axis=1)
future_regressors_df
```
Then during the forecast:

```
P_horizon = 576
future = m.make_future_dataframe(df = L12M, **regressors_df=future_regressors_df** , periods = P_horizon, n_historic_predictions=576) 
#def make_future_dataframe(self, df, events_df=None, regressors_df=None, periods=None, n_historic_predictions=False):

forecast1 = m.predict(df=future)
```

**Reply by ourownstory** [reply]
@Msaleh87 Thank you for following-up on this!
did it fail if you did not reset the index?
If so, please open an issue, we can drop the index internally instead.
